### ISSUE TITLE
Memory spike when using broker with batching policy

### ISSUE THREAD

**Issue by disintegrator** [Jeffail/benthos#997]
Hi, I'm trying to enable a batching policy that will allow me to generate 40MB CSVs from a large json lines (jsonl) file. While testing this I noticed that Benthos was getting OOM-killed consistently by docker as it exceed 2GB in memory usage. I managed to strip away my config to something that reproduces the issue and have captured it below. If I remove the broker that is around the drop output the pipeline runs successfully to completion. The issue also occurs I use a batching broker at the input instead of output.

---

## Command

```sh
docker run --rm --name readdrop -v $PWD/data:/data -v $PWD/project:/project jeffail/benthos:3.60.1 -c /project/config.yml
```

---

## Config

```yaml
input:
  file:
    paths: [/data/input/pypi/*.jsonl]
    codec: lines
  processors:
    - bloblang: |
        map escape_csv {
          root = if this.re_match("[\"\n,]+") {
            "\"" + this.replace("\"", "\"\"") + "\""
          } else {
            this
          }
        }

        meta = deleted()

        root = [
          (this.timestamp | "").string().apply("escape_csv"),
          (this.project | "").string().apply("escape_csv"),
          (this.url | "").string().apply("escape_csv"),
          (this.details.installer.name | "").string().apply("escape_csv"),
          (this.details.installer.version | "").string().apply("escape_csv"),
          (this.details.implementation.name | "").string().apply("escape_csv"),
          (this.details.implementation.version | "").string().apply("escape_csv"),
          (this.details.system.name | "").string().apply("escape_csv"),
          (this.details.system.release | "").string().apply("escape_csv")
        ].join(",")
output:
  broker:
    outputs:
      - drop: {}
    batching:
      period: 30s
      byte_size: 40_000_000
```

---

## Sample data

There is only one 942.2 MB file that is picked up by the file input and it is formatted as JSON lines. It's public data about PyPI requests that was available as a table in one of the public BigQuery datasets.

```jsonl
{"timestamp":"2021-11-29 00:41:50 UTC","country_code":"US","url":"/packages/e5/18/401594af67eda194a8b9167208621761927c937db7d60292608342bbac0a/pytest_cov-2.10.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.10.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.10.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.3.1"},"python":"3.9.9","implementation":{"name":"CPython","version":"3.9.9"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"59.3.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 15:02:12 UTC","country_code":"US","url":"/packages/3f/12/d5089675aa7955c9b7e8b831184b2b13ccece146c5b23250b0102a0d76e2/pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.0","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.3.1"},"python":"3.6.8","implementation":{"name":"CPython","version":"3.6.8"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"40.6.2"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 07:17:05 UTC","country_code":"US","url":"/packages/3f/12/d5089675aa7955c9b7e8b831184b2b13ccece146c5b23250b0102a0d76e2/pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.0","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"20.3.3"},"python":"3.8.12","implementation":{"name":"CPython","version":"3.8.12"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"58.3.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 18:41:37 UTC","country_code":"US","url":"/packages/3f/12/d5089675aa7955c9b7e8b831184b2b13ccece146c5b23250b0102a0d76e2/pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.0-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.0","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.0.1"},"python":"3.8.10","implementation":{"name":"CPython","version":"3.8.10"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"44.0.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 06:07:02 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.3.1"},"python":"3.7.12","implementation":{"name":"CPython","version":"3.7.12"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"47.1.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 13:10:42 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.3.1"},"python":"3.7.12","implementation":{"name":"CPython","version":"3.7.12"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"47.1.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 07:18:27 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"20.3.4"},"python":"2.7.13","implementation":{"name":"CPython","version":"2.7.13"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-aws"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"44.1.1"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 15:03:06 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.3.1"},"python":"3.8.12","implementation":{"name":"CPython","version":"3.8.12"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1021-azure"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"56.0.0"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 19:09:17 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.1.3"},"python":"3.7.11","implementation":{"name":"CPython","version":"3.7.11"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"5.11.0-1020-aws"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1l  24 Aug 2021","setuptools_version":"58.0.4"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
{"timestamp":"2021-11-29 01:58:17 UTC","country_code":"US","url":"/packages/e3/1a/6affecd2344efee7f2487fac82242474cbac09f9e04929da5944907baf11/pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","file":{"filename":"pytest_cov-2.11.1-py2.py3-none-any.whl","project":"pytest-cov","version":"2.11.1","type":"bdist_wheel"},"details":{"installer":{"name":"pip","version":"21.2.4"},"python":"3.6.15","implementation":{"name":"CPython","version":"3.6.15"},"distro":{"name":"Ubuntu","version":"20.04","id":"focal","libc":{"lib":"glibc","version":"2.31"}},"system":{"name":"Linux","release":"4.15.0-1115-aws"},"cpu":"x86_64","openssl_version":"OpenSSL 1.1.1f  31 Mar 2020","setuptools_version":"58.0.4"},"tls_protocol":"TLSv1.2","tls_cipher":"ECDHE-RSA-AES128-GCM-SHA256"}
```

---

## Environment

```
  System:
    OS: macOS 11.6.1
    CPU: (8) arm64 Apple M1
    Memory: 141.98 MB / 8.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.17 - /opt/homebrew/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
  Managers:
    Homebrew: 3.3.7 - /opt/homebrew/bin/brew
    pip3: 21.1.3 - ~/.pyenv/shims/pip3
    RubyGems: 3.2.22 - /Users/ghaidar/.rbenv/shims/gem
  Utilities:
    CMake: 3.21.4 - /opt/homebrew/bin/cmake
    Make: 3.81 - /usr/bin/make
    GCC: 4.2.1 - /usr/bin/gcc
    Git: 2.34.1 - /opt/homebrew/bin/git
    Clang: 12.0.5 - /usr/bin/clang
  Servers:
    Apache: 2.4.48 - /usr/sbin/apachectl
  Virtualization:
    Docker: 20.10.8 - /usr/local/bin/docker
  IDEs:
    Nano: 2.0.6 - /usr/bin/nano
    VSCode: 1.62.3 - /usr/local/bin/code
    Vim: 8.2 - /usr/bin/vim
    Xcode: /undefined - /usr/bin/xcodebuild
  Languages:
    Bash: 3.2.57 - /bin/bash
    Go: 1.17.3 - /opt/homebrew/bin/go
    Perl: 5.30.2 - /usr/bin/perl
    PHP: 7.3.29 - /usr/bin/php
    Python: 3.9.6 - /Users/ghaidar/.pyenv/shims/python
    Python3: 3.9.6 - /Users/ghaidar/.pyenv/shims/python3
    Ruby: 3.0.2 - /Users/ghaidar/.rbenv/shims/ruby
  Databases:
    PostgreSQL: 11.13 - /opt/homebrew/bin/postgres
    SQLite: 3.32.3 - /usr/bin/sqlite3
  Browsers:
    Chrome: 96.0.4664.55
    Firefox: 94.0.2
    Safari: 15.1
```

**Comment by Jeffail** [comment]
Hey @disintegrator there's no back pressure here but it's because of the `drop` output rather than the broker or batching policies. The reason you're only noticing it with the batching enabled is likely because that's the bit that's allowing your pipeline to hit a throughput high enough to fill memory, otherwise the lower rate of message traversal provides just enough back pressure to keep within limits.

In order to force some kind of back pressure you could add a rate limit processor at the output level: https://www.benthos.dev/docs/components/processors/rate_limit, it'd look something like this:

```yaml
output:
  broker:
    outputs:
      - drop: {}
        processors:
        - rate_limit:
            resource: drop_limit
    batching:
      period: 30s
      byte_size: 40_000_000

rate_limit_resources:
  - label: drop_limit
    local:
      count: 1000 # TODO: Tune this
      interval: 10ms
```

Or a simpler version would be to emulate a slow output with a sleep processor:

```yaml
output:
  broker:
    outputs:
      - drop: {}
        processors:
        - sleep:
            duration: 10ms
    batching:
      period: 30s
      byte_size: 40_000_000
```

We should probably add a natural rate limit internal to the drop output to avoid these issues but I'm not sure what that limit should be tuned for.

**Comment by disintegrator** [comment]
@Jeffail Amazing! This is a facepalm moment. I did not notice the rate limit processor 😅 

For what it's worth, the original config was writing the batches using the `file` output and the problem was the same. The config I shared is the minimum reproduction. I followed the docs advice of swapping bits out with `drop` to narrow down the issue.

**Comment by disintegrator** [comment]
@Jeffail I'm reopening this because I could not confirm that simulating a slow downstream output was causing backpressure. I tried an example from your advice and added a 1 hour sleep but my processor was quickly OOM-killed by docker:

```yaml
input:
  file:
    paths: [${DATA_PATH}/input/pypi/*.jsonl]
    codec: lines
  processors:
    - bloblang: |
        map escape_csv {
          root = if this.re_match("[\"\n,]+") {
            "\"" + this.replace("\"", "\"\"") + "\""
          } else {
            this
          }
        }

        meta = deleted()

        root = [
          (this.timestamp | "").string().apply("escape_csv"),
          (this.project | "").string().apply("escape_csv"),
          (this.url | "").string().apply("escape_csv"),
          (this.details.installer.name | "").string().apply("escape_csv"),
          (this.details.installer.version | "").string().apply("escape_csv"),
          (this.details.implementation.name | "").string().apply("escape_csv"),
          (this.details.implementation.version | "").string().apply("escape_csv"),
          (this.details.system.name | "").string().apply("escape_csv"),
          (this.details.system.release | "").string().apply("escape_csv")
        ].join(",")
output:
  broker:
    outputs:
      - drop: {}
        processors:
          - sleep:
              duration: 60m
    batching:
      period: 1m
      byte_size: 40_000_000
      processors:
        - insert_part:
            index: 0
            content: "timestamp,project,url,installer_name,installer_version,implementation_name,implementation_version,system_name,system_release"
        - archive:
            format: lines
        - bloblang: |
            meta batch_hash = content().hash("xxhash64").encode("hex")
```

I suspect the batcher goroutine(?) is not waiting/suspending while downstreams process messages. Perhaps that is exactly in line with what you mentioned in your original comment:

> The reason you're only noticing it with the batching enabled is likely because that's the bit that's allowing your pipeline to hit a throughput high enough to fill memory

If I understand this correctly then I think I'm making a feature request for a batching process that suspends until downstreams are ready in this issue. If that is the intended behaviour then I'll continue to investigate…

**Comment by Jeffail** [comment]
@disintegrator might be worth trying it with a reduced size as 40mb is very large relative to a 2gb limit, benthos messages are preserved internally in case of downstream errors so a 40mb batch will be buffered in potentially more than one form during its lifetime within the service. Each layer of components then adds one space for a message to be blocked waiting to progress, so I can imagine it's within the realm of possibility that you'd reach that limit even with back pressure working normally.

**Comment by disintegrator** [comment]
You're absolutely right. I just did a number of tests at different batch size with that 60m sleep and I was able to confirm that the memory held at stable point. Without the sleep my theory was that memory will fluctuate but not exceed the stable point and I am confident I confirmed this theory as well.

This all told me that backpressure was in effect and that I need to find what the stable point is for a 40MB batch size because it's not 2GB.

Thank you again for working with me on these questions!

**Comment by disintegrator** [comment]
@Jeffail You're absolutely right. I just did a number of tests at different batch size with that 60m sleep and I was able to confirm that the memory held at stable point. Without the sleep my theory was that memory will fluctuate but not exceed the stable point and I am confident I confirmed this theory as well.

This all told me that backpressure was in effect and that I need to find what the stable point is for a 40MB batch size because it's not 2GB.

Thank you again for working with me on these questions!
